source/common/websocket/codec.h

```diff
@@ -0,0 +1,133 @@
+
+#include "envoy/buffer/buffer.h"
+
+#include "source/common/common/logger.h"
+
+namespace Envoy {
+namespace WebSocket {
+
+// Opcodes (https://datatracker.ietf.org/doc/html/rfc6455#section-11.8)
+constexpr uint8_t kFrameOpcodeContinuation = 0;
+constexpr uint8_t kFrameOpcodeText = 1;
+constexpr uint8_t kFrameOpcodeBinary = 2;
+constexpr uint8_t kFrameOpcodeClose = 8;
+constexpr uint8_t kFrameOpcodePing = 9;
+constexpr uint8_t kFrameOpcodePong = 10;
+constexpr std::array<uint8_t, 6> kFrameOpcodes = {kFrameOpcodeContinuation, kFrameOpcodeText,
+                                                  kFrameOpcodeBinary,       kFrameOpcodeClose,
+                                                  kFrameOpcodePing,         kFrameOpcodePong};
+
+// Length of the masking key which is 4 bytes fixed size
+constexpr uint8_t kMaskingKeyLength = 4;
+// 16 bit payload length
+constexpr uint8_t kPayloadLength16Bit = 2;
+// 64 bit payload length
+constexpr uint8_t kPayloadLength64Bit = 8;
+
+// Wire format (https://datatracker.ietf.org/doc/html/rfc6455#section-5.2)
+// of WebSocket frame:
+//
+//   0                   1                   2                   3
+//   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//  +-+-+-+-+-------+-+-------------+-------------------------------+
+//  |F|R|R|R| opcode|M| Payload len |    Extended payload length    |
+//  |I|S|S|S|  (4)  |A|     (7)     |             (16/64)           |
+//  |N|V|V|V|       |S|             |   (if payload len==126/127)   |
+//  | |1|2|3|       |K|             |                               |
+//  +-+-+-+-+-------+-+-------------+ - - - - - - - - - - - - - - - +
+//  |     Extended payload length continued, if payload len == 127  |
+//  + - - - - - - - - - - - - - - - +-------------------------------+
+//  |                               | Masking-key, if MASK set to 1 |
+//  +-------------------------------+-------------------------------+
+//  | Masking-key (continued)       |          Payload Data         |
+//  +-------------------------------- - - - - - - - - - - - - - - - +
+//  : .... Payload Data continued .... Payload Data continued ..... :
+//  + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
+//  | .... Payload Data continued .... Payload Data continued ..... |
+//  +---------------------------------------------------------------+
+
+// In-memory representation of the contents of a WebSocket frame.
+struct Frame {
+  // Indicates that this is the final fragment in a message.
+  bool final_fragment_;
+  // Frame opcode.
+  uint8_t opcode_;
+  // The 4 byte fixed size masking key used to mask the payload. Masking/unmasking should be
+  // performed as described in https://datatracker.ietf.org/doc/html/rfc6455#section-5.3
+  absl::optional<uint32_t> masking_key_;
+  // Length of the payload as the number of bytes.
+  uint64_t payload_length_;
+  // WebSocket payload data (extension data and application data).
+  Buffer::InstancePtr payload_;
```

Comment:
Two questions:
* From the case of the issue https://github.com/envoyproxy/envoy/issues/13877#issuecomment-816508238, people may just want to parse the opcode, and don't care about the payload. And the length of payload can be huge (seems it can be 64bit), can we provide an option to the caller to not buffer the payload data, that will save a lot of memory. (since you make a copy of the payload to the struct Frame)
* Since the length of payload can be huge, should we enable option for the caller limits the length of payload to buffer.

Reply:
added another constructor with a bool to stop the payload copying to `Frame`